enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+
+We propose to use "green" and "red" scripts to decide if MicroShift's data 
+should be backed up (green) or restored (red). For reasons mentioned in section 
+"Alternatives - Performing backup on shutdown" it was concluded that both 
+backup and restore should happen on system start, rather than shutdown.
+It means that green or red script persists an action that will be performed 
+on next boot of the system.
+
+As a consequence, whether the next boot happens to be different or the same
+ostree deployment, it will produce a backup compatible with previously booted deployment
+and then attempt to perform a data migration if needed.
+It also means that consecutive red boots of new ostree deployment will restore the data,
+attempt to migrate it, and run MicroShift, i.e. each boot starts from the same place, just
+like it would be a first boot of that ostree deployment.
+This provides a safety net in case of invalid data migration - it will be attempted again,
+on each boot following red boot.
+
+Potential risk is possibility of losing data that might've been produced during window
+of MicroShift start and system reboot. However, only applies to MicroShift's data,
+because Kubernetes application's data isn't persisted in etcd.
+
+To integrate with greenboot, bash scripts will be placed in `/etc/greenboot/green.d` and
+`/etc/greenboot/red.d`.
+Depending on code complexity, they may include actual logic or just invoke 
+`microshift greenboot {green,red}` commands which would implement the logic.
+
+Both green and red scripts (or programs) will write to `/var/lib/microshift.bak/next_boot`.
+The file is outside of MicroShift data directory to reduce number of operations related to 
+making a backup by not needing to remove it.
+
+Green will persist following info:
+- "backup"
+- current ostree deployment id
+
+Red will persist following info:
+- "restore"
+
+### Backup and restore of MicroShift data
+
+To integrate fully with greenboot and ostree deployments MicroShift needs to be able to
+back up and restore its data.
+If new ostree deployment fails to be healthy and system is rolled back to previous
+deployment, MicroShift must also roll back in time to data compatible with 
+the older deployment.
+Because device administrator might want to manually go back to older deployment,
+it means that a backup for that particular deployment must be kept until it disappears
+from ostree. It results in requirement of keeping backup for each deployment separately
+regardless of the MicroShift version they feature.
+
+Because difference between two deployments might not be MicroShift itself, but applications
+that run on top of MicroShift, MicroShift's backups are tied to ostree deployments rather 
+than MicroShift versions.
+
+Decision whether to backup or restore is based on file persisted during previous boot
+(see "Integration with greenboot").
+That file will be removed after successful operation of backup or restore.
```

Comment:
It doesn't mean the file won't be there on shutdown.

1. system boots
2. greenboot starts healthchecks (microshift healthcheck runs in background)
3. pre-run
4. backup/restore is performed, file deleted
5. microshift starts
6. greenboot ok or nok, green/red -> file created

Reply:
You're right, the greenboot scripts would recreate the file.

What's the benefit of removing the file, though? Why not just overwrite it?

Reply:
To not let it gone stale in case of some error
- deployment boots,
- backup
- microshift starts but between now and green/red scripts system loses power
- system boots

Now, do we want to do backup again and potentially overwrite good data? 
Should we care? bad data could lead to rollback so device is healthy again
This is "restore" + missing backup (seems like we have many ways leading to this with potentially different actions)

Reply:
It could be better to be explicit. When MicroShift successfully starts, it could set the mode to "run". If we lose power, on restart we would bypass all of the upgrade steps. If greenboot fails, the mode would be set to "retry" or "restore" at that point.d

Reply:
I need to revisit workflows and even more details and consider further failure points like powerloss in various points in time